internal/pkg/stream/stream.go

```diff
@@ -64,18 +65,19 @@ func Stream(ctx context.Context, streamer Streamer) error {
 
 // nextFetchDate returns a time to wait using random jitter and exponential backoff.
 func nextFetchDate(clock clock, rand func(int) int, retries int) time.Time {
-	// waitMs := rand.Intn( 							// Get a random integer between 0 and ...
+	// waitMs := rand.Intn( 							// Get a random integer between streamerMinFetchIntervalDurationMs and ...
 	// 	min( 											// the minimum of ...
 	// 		streamerMaxFetchIntervalDuration,           // the max fetch interval and ...
 	// 		streamerFetchIntervalDuration*(1<<retries), // d*2^r, where r=retries and d= the normal
-	// 	),
-	// )
-	waitMs := rand(
-		min(
-			streamerMaxFetchIntervalDurationMs,
-			streamerFetchIntervalDurationMs*(1<<retries),
-		),
-	)
+	// 	)-streamerMinFetchIntervalDurationMs
+	// ) + streamerMinFetchIntervalDurationMs
+	waitMs :=
+		rand(
+			min(
+				streamerMaxFetchIntervalDurationMs,
+				streamerFetchIntervalDurationMs*(1<<retries),
+			)-streamerMinFetchIntervalDurationMs,
+		) + streamerMinFetchIntervalDurationMs
```

Comment:
while i think, in practice, this will always work - what we are doing here is effectively hoping that we'll give the scheduler enough time to give the `select` block a chance to run again and select the `Done()` case before the time returned has passed, which isn't necessarily guaranteed.

i wonder if it makes more sense to handle the "closed" case from within the `Fetch()` functions. for example, we could change the logic in `Fetch()` to something like:
```
Fetch() {
  if we closed the chan {
    return time.Time{}
  }

  return nextFetchDate()
}
```

And then in the `select` block, we could change the fetch delay case to:
```
case <-time.After(fetchDelay):
	next, err = streamer.Fetch()
	if err != nil {
		return err
	}
	streamer.Notify()
    if next.IsZero() {
		return nil
    }
```
Something like that? We could also return a "Stop fetching" error or something like that, or a third return value. Something other than just a time that should be long enough.

Reply:
That sounds good to me. I'll still leave the min fetch time so that we don't fetch too often, because right now it is almost guaranteed that we'll receive the throttling error.